### PR TITLE
Nv2023042 various jira tickets

### PR DESCRIPTION
**Pull request by nvolk**
Two new validators/fixers:

- update-field-540: Fix MELKEHITYS-2431 for updating field 540 contents
- sanitize-vocabulary-source-codes: MRA-532: sanitize vocabulary source codes

### CODE REVIEW

**Comment by nvolk** [comment]
Well, I added fix for MET-336 here as well.

File: src/sanitize-vocabulary-source-codes.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+// Note that https://github.com/NatLibFi/marc-record-validators-melinda/blob/master/src/unicode-decomposition.js contains
+// similar functionalities. It's less generic and lacks diacritic removal but has it advantages as well.
+
+//const debug = createDebugLogger('@natlibfi/melinda-marc-record-merge-reducers/reducers/normalize-utf-diacritics');
+
+// See also https://github.com/NatLibFi/marc-record-validators-melinda/blob/master/src/unicode-decomposition.js .
+// It uses a list of convertable characters whilst this uses a generic stuff as well.
+// It handles various '.' and '©' type normalizations as well.
+// NB! This version has minor bug/feature issue regarding fixComposition()
+
```
Comment (ammsalme) [inline src/sanitize-vocabulary-source-codes.js:13]: Are these valid comments for this validator?

File: src/sanitize-vocabulary-source-codes.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  }
+  // As we can't fix this here, apply this yso-rule only when validating!
+  if (!fix && subfield.value.indexOf('yso/') === 0) {
+    return !['yso/eng', 'yso/fin', 'yso/swe'].includes(subfield.value);
+  }
+  return false;
+}
+
+function fieldHasCrappySubfield2(field, fix) {
+  if (!field.tag.match(/^(?:6..|257|38.)$/u)) {
```
Comment (ammsalme) [inline src/sanitize-vocabulary-source-codes.js:87]: Add 370

File: src/sanitize-vocabulary-source-codes.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    return false;
+  }
+  return field.subfields.some(sf => isCrappySubfield2(sf, fix));
+}
+
+function getFieldsWithCrappySubfieldCode(record, fix) {
+  // NB! $2 is used in fields 257, 38X etc. However, these were not requested, so I'll skip them here in the first version.
+  return record.get(/^6..$/u).filter(f => fieldHasCrappySubfield2(f, fix));
```
Comment (ammsalme) [inline src/sanitize-vocabulary-source-codes.js:95]: We could validate/fix also 257, 370, 38X

File: src/sanitize-vocabulary-source-codes.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    if (fix) {
+      subfield.value = fixedVersion; // eslint-disable-line functional/immutable-data
+      return true;
+    }
+
+    return true;
+  }
+  // As we can't fix this here, apply this yso-rule only when validating!
+  if (!fix && subfield.value.indexOf('yso/') === 0) {
+    return !['yso/eng', 'yso/fin', 'yso/swe'].includes(subfield.value);
```
Comment (ammsalme) [inline src/sanitize-vocabulary-source-codes.js:81]: We could have language checks also for slm and mts.

File: src/update-field-540.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+// Note that https://github.com/NatLibFi/marc-record-validators-melinda/blob/master/src/unicode-decomposition.js contains
+// similar functionalities. It's less generic and lacks diacritic removal but has it advantages as well.
+
+//const debug = createDebugLogger('@natlibfi/melinda-marc-record-merge-reducers/reducers/normalize-utf-diacritics');
+
+// See also https://github.com/NatLibFi/marc-record-validators-melinda/blob/master/src/unicode-decomposition.js .
+// It uses a list of convertable characters whilst this uses a generic stuff as well.
+// It handles various '.' and '©' type normalizations as well.
+// NB! This version has minor bug/feature issue regarding fixComposition()
+
```
Comment (ammsalme) [inline src/update-field-540.js:13]: Comments for some other validator?

File: src/update-field-540.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    const subfieldsC = field.subfields.filter(sf => validLicenseInSubfieldC(sf, validLicense));
+    subfieldsC.forEach(c => fixC(field, c));
+  }
+
+  return true;
+}
+
+
+function getFieldsThatUseOldFormat(record, fix) {
+  // NB! $2 is used in fields 257, 38X etc. However, these were not requested, so I'll skip them here in the first version.
```
Comment (ammsalme) [inline src/update-field-540.js:104]: Old comment, remove.

**Review by ammsalme — APPROVED** [review]

**Pull request merged**